Refactor jukebox sal to use models

AhmedHanafy725 opened the pull request:
### Description

- Refactor jukebox sal to use models.
- Clean up the code.
- Add a delete button to the nodes in the deployment.
- Remove ssh key from deployment chatflow.
- Add update deployment instance service.
- Add self-healing service.
- Locking deployment model while writing.
- Fix listing by deployment type.
- Delete deployment in case of failure.

### Related Issues
#10
#45
#46
#47
#48
#49

ashraffouda committed:
Adds models

Signed-off-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

AhmedHanafy725 committed:
Refactor using models

ranatrk committed:
Update list deployments and delete deployment and fund wallet in bottle server and UI

ranatrk committed:
Add node deletion

Fix auto extend button frontend

ranatrk committed:
Update monitoring service to use deployment models

AhmedHanafy725 committed:
Fix activeworkloads count in frontend

ashraffouda committed:
Adds new flist for digibyte and fixes sending email

Signed-off-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

AhmedHanafy725 committed:
Remove ssh key from deployment chatflow

Add update deployment service

AhmedHanafy725 committed:
Split utils functions in a new file

Add locking for writing on deployment object

AhmedHanafy725 committed:
Remove jukebox deployment instance in case of failure

AhmedHanafy725 committed:
Fix list_deployment with more than one query

AhmedHanafy725 committed:
WIP: Add self healing service

waleedhammam commented on jumpscale/packages/jukebox/chats/digibyte.py:
```diff
@@ -24,14 +23,12 @@ class DigibyteDeploy(JukeboxDeployChatflow):
-            "pub_key": self.public_key,
-            }
+        self.env = {}
+        # self.env = {
+        #     "pub_key": self.public_key,
+        # }
```
```suggestion
```

waleedhammam commented on jumpscale/packages/jukebox/chats/ubuntu.py:
```diff
@@ -25,13 +24,10 @@ class UbuntuDeploy(JukeboxDeployChatflow):
     def environment(self):
-        self.env = {"pub_key": self.public_key}
+        self.env = {}
+        # self.env = {"pub_key": self.public_key}
```
```suggestion
```

waleedhammam commented on jumpscale/packages/jukebox/serivces/monitor.py:
```diff
@@ -16,42 +16,46 @@ def __init__(self, interval=60 * 60 * 12, *args, **kwargs):
+    def _check_down_containers(self, deployment):
+        user = j.data.text.removeprefix(deployment.identity_name, "jukebox_")
+
+        # for deployment_type, deployments in all_deployments.items():
+        #     for deployment in deployments:
```
```suggestion
```

waleedhammam commented on jumpscale/packages/jukebox/serivces/monitor.py:
```diff
@@ -16,42 +16,46 @@ def __init__(self, interval=60 * 60 * 12, *args, **kwargs):
+            )
+            message = (
+                f"Dear {user},\n\n"
+                f"{number_of_down_deployment} node(s) of {deployment.solution_type} went down for your deployment {deployment.deployment_name}."
```
```suggestion
                f"{number_of_down_deployment} node(s) of {deployment.solution_type} went down for your deployment {deployment.deployment_name}. The system will try to bring it back, Please refer for your jukebox dashboard for more information."
```

waleedhammam commented on jumpscale/sals/jukebox/__init__.py:
```diff
@@ -0,0 +1,66 @@
+            deployment.save()
+
+
+BCNodeFACTORY = BlockchainStoredFactory(JukeboxDeployment)
+BCNodeFACTORY.always_reload = True
```
better to be at the top of the file ?

AhmedHanafy725 replied:
no, it needs to be here after the class definition

waleedhammam commented on jumpscale/sals/jukebox/jukebox.py:
```diff
@@ -0,0 +1,324 @@
+        )
+
+        if result:
+            # self.md_show_update("Deploying Network on Nodes....")
```
```suggestion
```

waleedhammam commented on jumpscale/sals/jukebox/jukebox.py:
```diff
@@ -0,0 +1,324 @@
+            env=env,
+            interactive=False,
+            entrypoint=entry_point,
+            # log_config=self.log_config,
```
```suggestion
```

waleedhammam commented on jumpscale/sals/jukebox/jukebox_deploy_chatflow.py:
```diff
@@ -1,31 +1,48 @@
         "deploy",
         "success",
     ]
     # FLIST = "<flist_url>"
```
```suggestion
```

AhmedHanafy725 replied:
We need this for anyone adding a new child chatflow

waleedhammam commented on jumpscale/sals/jukebox/utils.py:
```diff
@@ -0,0 +1,140 @@
+        cru=cru * number_of_deployments,
+        mru=mru * number_of_deployments,
+        sru=sru * number_of_deployments,
+        # ip_version=None,
+        # no_nodes=1,
```
```suggestion
```

AhmedHanafy725 committed:
Add default value to deployment's expiration

Add node to block nodes in case of any failure in network workload

Fix removing error nodes after redeploy them

AhmedHanafy725 committed:
Apply suggestions from code review

Co-authored-by: Waleed <waleed.hammam@gmail.com>

AhmedHanafy725 committed:
Merge branch 'jukemodels' of github.com:threefoldtech/tf_jukebox into jukemodels

AhmedHanafy725 committed:
add comment for flist commented line
